### PR TITLE
provider-keymgmt(7): Document the KEYMGMT interface

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by levitte** [comment]
Now that #9409 has been merged, this one is ready for review.

File: doc/man7/provider-keymgmt.pod [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+OP_keymgmt_genkey() should generate keys and create a provider side
+structure for them.
+Values from the passed domain parameters I<domparams> as well as from
+the passed B<OSSL_PARAM> array I<params> should be used as input for
+key generation.
+
+OP_keymgmt_loadkey() should return a provider side key structure with
+a key loaded from a location known only to the provider, identitified
+with the identity I<id> or size I<idlen>.
```
Comment (paulidale) [inline doc/man7/provider-keymgmt.pod]: s/or/and/

**Review by paulidale — CHANGES_REQUESTED** [review]
I wonder about the prevalence of _should_ putting the function descriptions into future tense.

File: doc/man7/provider-keymgmt.pod [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ int OP_keymgmt_exportdomparams(void *domparams, OSSL_PARAM params[]);
+
+ /* Key domain parameter discovery */
+ const OSSL_PARAM *OP_keymgmt_importdomparam_types(void);
+ const OSSL_PARAM *OP_keymgmt_exportdomparam_types(void);
+
+ /* Key creation and destruction */
+ void *OP_keymgmt_importkey(void *provctx, const OSSL_PARAM params[]);
+ void *OP_keymgmt_genkey(void *provctx,
+                      void *domparams, const OSSL_PARAM genkeyparams[]);
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: Fixed

File: doc/man7/provider-keymgmt.pod [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ int OP_keymgmt_exportkey(void *key, OSSL_PARAM params[]);
+
+ /* Key discovery */
+ const OSSL_PARAM *OP_keymgmt_importkey_types(void);
+ const OSSL_PARAM *OP_keymgmt_exportkey_types(void);
+
+=head1 DESCRIPTION
+
+The KEYMGMT operation doesn't have much public visibility in OpenSSL
+libraries, it's a rather internal one that's designed to work in
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: Edited, but not quite like you suggested.

File: doc/man7/provider-keymgmt.pod [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+tandem with operations that use private/public key pairs.
+
+Because the KEYMGMT operation share knowledge with the operations it
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: You're right

File: doc/man7/provider-keymgmt.pod [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+works in tandem with, they must belong to the same provider.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: Okie

File: doc/man7/provider-keymgmt.pod [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+The OpenSSL libraries will ensure that they do.
+
+The primary responsibility of the KEYMGMT operation is to hold the
+provider side domain parameters and keys for the OpenSSL library
+EVP_PKEY structure.
+
+All "functions" mentioned here are passed as function pointers between
+F<libcrypto> and the provider in B<OSSL_DISPATCH> arrays via
+B<OSSL_ALGORITHM> arrays that are returned by the providers
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: Thanks

File: doc/man7/provider-keymgmt.pod [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+provider_query_operation() function
+(see L<provider-base(7)/Provider Functions>).
+
+All these "functions" have a corresponding function type definition
+named B<OSSL_{name}_fn>, and a helper function to retrieve the
+function pointer from a B<OSSL_DISPATCH> element named
+B<OSSL_get_{name}>.
+For example, the "function" OP_keymgmt_importdomparams() have these:
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: Thanks

File: doc/man7/provider-keymgmt.pod (hunk 1 above)
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: I meant to write "of", not "or"...  the size doesn't identify the key.

Comment (paulidale) [reply doc/man7/provider-keymgmt.pod]: _of_ is fine too.  Both work here.

File: doc/man7/provider-keymgmt.pod (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply doc/man7/provider-keymgmt.pod]: I thought apostrophes were to be avoided in documentation :)

**Review by paulidale — APPROVED** [review]
Looks okay now.

File: doc/man7/provider-keymgmt.pod (hunk 3 above)
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-keymgmt.pod]: It's been *discussed*, but we haven't made an actual decision on this.  I doubt we ever will.

**Comment by levitte** [comment]
Thank you for the review!

Merged.

da2addc515d547b0d724a4fc730c4345ed713221 provider-keymgmt(7): Document the KEYMGMT interface

**Pull request closed without merging**